### PR TITLE
Add cross-compile and multi-arch push support

### PR DESCRIPTION
**Pull request by deitch**
## Description

Adds the following:

* cross-compile support from `amd64` to all supported arches
* standard Makefile targets `build` and `image` which support `ARCH`
* standard Makefile target `image-all` to create all supported arches
* standard Makefile target `push` to push out a given `IMAGETAG` for an individual `ARCH`
* standard Makefile target `push-all` to push out a given `IMAGETAG` for all supported arches
* standardizes the `Dockerfile`s and simplifies them

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Standard support for cross-compile build, image and push for all supported arches
```

### CODE REVIEW

File: Dockerfile.amd64 [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM amd64/busybox:1.28.3
+
+LABEL maintainer "Tom Denham <tom@tigera.io>"
+
+ADD dist/amd64/ /opt/cni/bin/
```
Comment (fasaxc) [inline Dockerfile.amd64]: I'm _slightly_ wary of this sort of change.  Previously, we whitelisted the files that went into the container, now we'll include anything that happens to be in that directory.  The thing that usually causes this to go wrong is when you switch branches without running `make clean` and the list of files in the directory is different on the different branches.

That said, the release target does an explicit clean, so we're fairly safe from _shipping_ unexpected files. I think I can live with it given it's a nice simplification.

File: Makefile [hunk 2]
```diff
@@ -1,23 +1,38 @@
...
-	GO_BUILD_VER:=latest
+# unless otherwise set, I am building for my own architecture, i.e. not cross-compiling
+ARCH ?= $(BUILDARCH)
+
+# canonicalized names for target architecture
+ifeq ($(ARCH),aarch64)
+        override ARCH=arm64
+endif
+ifeq ($(ARCH),x86_64)
+    override ARCH=amd64
```
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: Nice catch. Amending.

File: Makefile [hunk 3]
```diff
@@ -41,32 +56,67 @@ CNI_SPEC_VERSION?=0.3.1
...
-# Release targets 
+# push images of any tag
+###############################################################################
+imagetag:
+ifndef IMAGETAG
+	$(error IMAGETAG is undefined - run using make <target> IMAGETAG=X.Y.Z)
+endif
+
+push: image imagetag
+	docker tag $(CONTAINER_NAME):latest-$(ARCH) $(CONTAINER_NAME):$(IMAGETAG)-$(ARCH)
```
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: OK. Well, especially if you two agree, and I didn't exactly love it in the first place, I am not going to argue too strenuously. :-)

So the behaviour of `push` is just to push out images whose name is `$(CONTAINER_NAME):$(IMAGE_TAG)-$(ARCH)` (and if `amd64`, arch-less as well), but not to tag them. It then will assume the tagging was done, in this case in `release-build`.

If I understand you correctly, the problem isn't _having_ a target that does then tagging, but the fact that _push_ always tags `latest` to `IMAGETAG` and thus assumes that `latest` is the one we want going out. The solution, as you both pointed out, is to have a `push` push-only target, and a tagging-only target, so we can invoke separately (push shouldn't _have_ to imply tag, as you both pointed out). 

Looks like `release-build` **is** build+tagging-only anyways. so I will create that, which should make the process a bit easier without inducing the risk. Let me know with the latest here after I push it out.

**Comment by deitch** [comment]
Phew. Rebase complete. Back to you @fasaxc

**Pull request merged**